File: org.amanzi.awe.report/ruby/gis.rb
require 'java'

puts "trying to include GIS"

include_class 'org.amanzi.awe.report.util.GIS'
include_class 'net.refractions.udig.project.ui.ApplicationGIS'
include_class 'org.amanzi.awe.neostyle.NeoStyleContent'
include_class 'org.amanzi.awe.filters.experimental.GroupFilter'

#include_class 'org.amanzi.awe.catalog.neo.GeoNeo'

include_class 'org.geotools.feature.DefaultFeature'
include_class 'org.geotools.data.FeatureSource'
include_class 'org.geotools.feature.FeatureCollection'
include_class 'com.vividsolutions.jts.geom.Geometry'

puts "GIS included"
include_class 'net.refractions.udig.project.internal.impl.LayerImpl'
include_class 'net.refractions.udig.project.internal.SimpleBlackboard'

class DefaultFeature
  def geometry
    getDefaultGeometry()
  end
end

class Geometry
  def ==(other)
    self.getGeometryType().upcase==other.to_s.upcase
  end
end

class SimpleBlackboard
  def filter(property=nil,&block)
    f=GroupFilter.new(property)
    f.setup(&block)
    put("FILTER",f)
  end
end

class LayerImpl
  def style=(style)
    neo_style=getStyleBlackboard().get(NeoStyleContent::ID)
    neo_style.clearStyle()
    style.styles.each do |shape,color_scheme|
      neo_style.addStyle(shape,color_scheme.colors.to_java(java.awt.Color))
    end
  end

  def geo_filter=(feature)
    getBlackboard().put("GEO_FILTER",feature)
  end
  
  def group_filter(property=nil,&block)
    puts "LayerImpl.group_filter"
    puts block_given?
    gfilter=GroupFilter.new(property)
    gfilter.setup(&block)
    #    getBlackboard().put("FILTER",[filter].to_java(GroupFilter))
    gfilter
  end

  def filters(&block)
    puts "LayerImpl.filters"
    self.instance_eval &block
    getBlackboard().put("FILTER",@filters.to_java(GroupFilter))
  end

  def add(filter)
    puts "LayerImpl.add"
    if @filters.nil?
      @filters=[filter]
    else
      @filters<<filter
    end
  end

  def filter=(filter)
    puts "LayerImpl.filter="
    getBlackboard().put("FILTER",[filter].to_java(GroupFilter))
  end

  #  def blackboard
  #    getBlackboard()
  #  end
end

include_class 'net.refractions.udig.project.internal.impl.MapImpl'

class MapImpl
  def copy
    ApplicationGIS::copyMap(self)
  end

  def layers
    LayerIterator.new(self.getMapLayers())
  end

  def print_features
    GIS::printFeatures(self)
  end

  def selected_features
    features=[]
    getMapLayers().each do |layer|
      source=layer.getResource(FeatureSource.java_class,nil);
      if !source.nil?
        collection = source.getFeatures(layer.getFilter())
        if !collection.nil?
          featuresIterator = collection.features()
          begin
            while featuresIterator.hasNext()
              features<<featuresIterator.next()
            end
          ensure
            featuresIterator.close()
          end
        end
      end
    end
    features
  end
end

class MapIterator
  def initialize(collection)
    #    puts "initialize"
    @collection=collection
  end

  def each
    iter=iterator
    while (iter.hasNext)do
      yield iter.next
    end
  end

  def first
    iter=iterator
    iter.hasNext ? iter.next : ApplicationGIS::NO_MAP
  end

  def iterator
    @collection.iterator
  end
end

class LayerIterator
  def initialize(collection)
    @collection=collection
  end

  def each
    iter=iterator
    while (iter.hasNext)do
      yield iter.next
    end
  end

  def iterator
    @collection.iterator
  end

  def find(hash,&block)
    type=hash[:type]
    iter=iterator
    layers=[]
    while (iter.hasNext)do
      layer=iter.next
      resource = layer.findGeoResource($geo_neo_class)
      if !resource.nil?
        geo_neo=resource.resolve($geo_neo_class,nil)
        layers<<layer if geo_neo.getGisType().to_s==type.to_s.upcase
      end
    end
    layers
  end
end

class GIS
  #Returns all maps
  def self.maps
    MapIterator.new(GIS::getAllMaps())
  end

end

File: org.amanzi.awe.report/src/org/amanzi/awe/report/util/GIS.java
/* AWE - Amanzi Wireless Explorer
 * http://awe.amanzi.org
 * (C) 2008-2009, AmanziTel AB
 *
 * This library is provided under the terms of the Eclipse Public License
 * as described at http://www.eclipse.org/legal/epl-v10.html. Any use,
 * reproduction or distribution of the library constitutes recipient's
 * acceptance of this agreement.
 *
 * This library is distributed WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */

package org.amanzi.awe.report.util;

import java.io.IOException;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.amanzi.awe.catalog.neo.GeoNeo;
import org.amanzi.awe.filters.experimental.GroupFilter;
import org.eclipse.core.runtime.NullProgressMonitor;
import org.eclipse.jface.dialogs.MessageDialog;
import org.geotools.data.FeatureSource;
import org.geotools.feature.Feature;
import org.geotools.feature.FeatureCollection;
import org.geotools.feature.FeatureIterator;

import com.vividsolutions.jts.geom.Coordinate;
import com.vividsolutions.jts.geom.Envelope;
import com.vividsolutions.jts.geom.Geometry;
import com.vividsolutions.jts.geom.MultiPolygon;

import net.refractions.udig.catalog.IGeoResource;
import net.refractions.udig.project.ILayer;
import net.refractions.udig.project.IMap;
import net.refractions.udig.project.IProjectElement;
import net.refractions.udig.project.internal.Blackboard;
import net.refractions.udig.project.internal.ProjectPlugin;
import net.refractions.udig.project.internal.impl.LayerImpl;
import net.refractions.udig.project.internal.impl.MapImpl;
import net.refractions.udig.project.ui.ApplicationGIS;

/**
 * TODO Purpose of
 * <p>
 * </p>
 * 
 * @author Pechko_E
 * @since 1.0.0
 */
public class GIS {
    public static final String FILTER = "FILTER";

    public static IGeoResource findGeoNeo(ILayer layer) {
        return layer.findGeoResource(GeoNeo.class);
    }

    public static GeoNeo resolveGeoNeo(ILayer layer) throws IOException {
        final long start = System.currentTimeMillis();
        final IGeoResource resource = layer.findGeoResource(GeoNeo.class);
        final GeoNeo geoNeo = resource.resolve(GeoNeo.class, null);
        System.out.println("[DEBUG] time: " + (System.currentTimeMillis() - start));
        return geoNeo;
    }

    public static List<MapImpl> getAllMaps() {
        List<MapImpl> maps = new ArrayList<MapImpl>();
        final List<IProjectElement> elements = ProjectPlugin.getPlugin().getProjectRegistry().getCurrentProject().getElements();
        for (IProjectElement element : elements) {
            if (element instanceof MapImpl) {
                MapImpl map = (MapImpl)element;
                if (!map.eIsProxy()) {
                    maps.add(map);
                }
            }
        }
        return maps;
    }

    public static void printFeatures(IMap map) {
        for (ILayer layer : map.getMapLayers()) {
            try {
                FeatureSource source = layer.getResource(FeatureSource.class, new NullProgressMonitor());
                if (source != null) {
                    FeatureCollection collection = source.getFeatures(layer.getFilter());

                    if (collection != null) {
                        FeatureIterator features = collection.features();

                        try {
                            while (features.hasNext()) {
                                final Feature feature = features.next();
                                System.out.println("[DEBUG] feature " + (feature == null ? null : feature));
                                final Geometry defaultGeometry = feature.getDefaultGeometry();
                                if (defaultGeometry!=null && defaultGeometry instanceof MultiPolygon){
                                    System.out.println("[DEBUG] Geometry type=" + defaultGeometry.getGeometryType()+" "+(defaultGeometry == null ? null : defaultGeometry));
                                    final Envelope envelope = defaultGeometry.getEnvelopeInternal();
                                    System.out.println("[DEBUG] envelope " + envelope);
                                    System.out.println("[DEBUG] coordinates:");
                                    for (Coordinate c:defaultGeometry.getCoordinates()){
                                        System.out.println(c.x+"\t"+c.y);
                                    }
                                    
                                }
                                
                            }
                        } finally {
                            features.close();
                        }
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
    public static  List<Feature> getSelectedFeatures(IMap map) {
        List<Feature> features=new ArrayList<Feature>();
        for (ILayer layer : map.getMapLayers()) {
            try {
                FeatureSource source = layer.getResource(FeatureSource.class, new NullProgressMonitor());
                if (source != null) {
                    FeatureCollection collection = source.getFeatures(layer.getFilter());
                    
                    if (collection != null) {
                        FeatureIterator featuresIterator = collection.features();
                        
                        try {
                            while (featuresIterator.hasNext()) {
                                features.add(featuresIterator.next());
                            }
                        } finally {
                            featuresIterator.close();
                        }
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return features;
    }
}
